### ISSUE TITLE
[Roadmap] False Positive Rate

### ISSUE THREAD

**Issue by flip1995** [rust-lang/rust-clippy#6623]
In the worst case, new lints are only available in nightly for 2 weeks, before
hitting beta and ultimately stable. This and the fact that fewer people use
nightly Rust nowadays makes it more probable that a lint with many FPs hits
stable. This leads to annoyed users, that will disable these new lints in the
best case and to more annoyed users, that will stop using Clippy in the worst.
A process should be developed and implemented to prevent this from happening.

#6429

Steps to completion:
- [x] Implement crater aka `cargo dev-lintcheck` for Clippy (https://github.com/rust-lang/rust-clippy/pull/6469)
- [ ] Implement framework so that lints can be enabled in nightly only
    - Previous attempt to implement this in Clippy #8211
- [x] Define clear process for stabilizing lints
    - We now have a FCP process where 2 team members have to approve on a new lint and no other team member vetos it.
- [ ] ~~CFP: Ask people using Clippy to submit their crates, so we can check new lints on them before releasing the lints to stable~~
    - We test ~200 crates on CI, so that wouldn't help much more.
- [x] Make it easy to run the `crater` tool on peoples private code / CI to get even more input
    - Lintcheck is now run on CI

**Comment by EdJoPaTo** [comment]
Something I think that would be helpful to deal with false positives would be #3122. In case false positives are found they will be allowed short term.
In the long term this false positive will be fixed and the allow comment shouldn't be there anymore.

**Comment by flip1995** [comment]
Good news: @xFrednet is currently working on this: https://github.com/rust-lang/rust/pull/87835

**Comment by xFrednet** [comment]
I would like to take a stab at the second point: *"Implement framework so that lints can be enabled in nightly only"*. I have three ideas how this can be implemented, with advantages and drawbacks. I'll create an issue outlining the possibilities and to ask for feedback. :upside_down_face:

**Comment by matthiaskrgr** [comment]
To add an additional datapoint, I have run clippy on all files (standalone) in the rustc repo [ [1]](github.com/matthiaskrgr/icemaker/)
(basically: run cargo clippy --fix -Wall::lints on a file, check if rustfix fails to apply clippys suggestion, bisect the offending lint), and these are the lints and the number of times a applied lint suggestion caused a compilation failure (aka invalid suggestion)

````
      1       "--force-warn clippy::assign-op-pattern",
      1       "--force-warn clippy::blocks-in-if-conditions",
      1       "--force-warn clippy::bool-comparison",
      1       "--force-warn clippy::bool-to-int-with-if",
      1       "--force-warn clippy::borrow-deref-ref",
      1       "--force-warn clippy::cloned-instead-of-copied",
      1       "--force-warn clippy::derive-partial-eq-without-eq",
      1       "--force-warn clippy::excessive-precision",
      1       "--force-warn clippy::explicit-auto-deref",
      1       "--force-warn clippy::match-single-binding",
      1       "--force-warn clippy::needless-arbitrary-self-type",
      1       "--force-warn clippy::needless-for-each",
      1       "--force-warn clippy::needless-late-init",
      1       "--force-warn clippy::range-minus-one",
      1       "--force-warn clippy::range-plus-one",
      1       "--force-warn clippy::redundant-pattern-matching",
      1       "--force-warn clippy::ref-binding-to-reference",
      1       "--force-warn clippy::short-circuit-statement",
      1       "--force-warn clippy::single-component-path-imports",
      1       "--force-warn clippy::transmutes-expressible-as-ptr-casts",
      1       "--force-warn clippy::unnecessary-sort-by",
      2       "--force-warn clippy::cast-lossless",
      2       "--force-warn clippy::equatable-if-let",
      2       "--force-warn clippy::manual-async-fn",
      2       "--force-warn clippy::needless-lifetimes",
      2       "--force-warn clippy::redundant-closure",
      2       "--force-warn clippy::redundant-closure-call",
      2       "--force-warn clippy::redundant-pub-crate",
      2       "--force-warn clippy::unused-unit",
      3       "--force-warn clippy::needless-return",
      3       "--force-warn clippy::no-mangle-with-rust-abi",
      3       "--force-warn clippy::redundant-clone",
      3       "--force-warn clippy::trait-duplication-in-bounds",
      3       "--force-warn clippy::unnecessary-fold",
      3       "--force-warn clippy::unnested-or-patterns",
      4       "--force-warn clippy::unit-arg",
      4       "--force-warn clippy::unnecessary-operation",
      4       "--force-warn clippy::wildcard-imports",
      6       "--force-warn clippy::redundant-closure-for-method-calls",
      6       "--force-warn clippy::use-self",
     13       "--force-warn clippy::ptr-as-ptr",
     23       "--force-warn clippy::let-unit-value",
     40       "--force-warn clippy::borrow-as-ptr",
 `````
Note that this is by no means a complete picture, it only shows lints that can be applied automatically in the first place, and lints that are more likely to fire on code inside short snippets of rustcs unit tests.

**Comment by llogiq** [comment]
I wonder how many of those are in compile tests that would have failed to compile in the first place?

Do you have the raw data somewhere?

**Comment by matthiaskrgr** [comment]
I think none of those failed to compile in the first place because I don't pass `--broken code` flag to rustfix which is required to fix code that does not compile in the first place iirc..? I can't check in detail right now.

**Comment by matthiaskrgr** [comment]
I uploaded the errors.json which contains the files and the clippy lints that failed to apply with them.
The summary was obtained via `cat errors.json | grep clippy:: | sort -n | uniq -c | sort -n` 
[errors.json.tar.gz](https://github.com/rust-lang/rust-clippy/files/10932327/errors.json.tar.gz)

**Comment by xFrednet** [comment]
> Implement framework so that lints can be enabled in nightly only

I've tried to implement this in Clippy, but we determined, that this should be implemented in rustc instead. The compiler team requested us to create a [MCP](https://forge.rust-lang.org/compiler/mcp.html) if we want this for rustc.

@Jarcho created a working implementation in rustc, but it was closed until an MCP was created and accepted: https://github.com/rust-lang/rust/pull/109063.

**Comment by lolbinarycat** [comment]
@[EdJoPaTo](https://github.com/EdJoPaTo)
[on Nov 4, 2021](https://github.com/rust-lang/rust-clippy/issues/6623#issuecomment-960617276)

> Something I think that would be helpful to deal with false positives would be https://github.com/rust-lang/rust-clippy/issues/3122. In case false positives are found they will be allowed short term.
In the long term this false positive will be fixed and the allow comment shouldn't be there anymore

the newly stabalized `expect` attribute should also help with this.

**Comment by flip1995** [comment]
I will close this issue, as the only real open point is to work out if we want nightly lints. But that is its own whole project (that needs to be done in rustc).

So we done it! Another point checked of the 2021 roadmap 😁
